### ISSUE TITLE
Consider adding `np.interp` analog

### ISSUE THREAD

**Issue by ev-br** [data-apis/array-api-extra#193]
suggested at https://github.com/scipy/scipy/pull/22343#discussion_r2020202353

**Comment by lucascolley** [comment]
first bit of due diligence: https://github.com/data-apis/array-api/issues?q=interp is blank

**Comment by rgommers** [comment]
That definitely won't fly as part of the standard.

**Comment by lucascolley** [comment]
The NumPy implementation is in C at https://github.com/numpy/numpy/blob/c23886c226012a50da11066d3be98fd94e571101/numpy/_core/src/multiarray/compiled_base.c#L552-L719

**Comment by ev-br** [comment]
Linear interpolation is certainly possible to write in pure python, here's one implementation (most likely, quite overengineered):  https://github.com/scipy/scipy/blob/v1.15.2/scipy/interpolate/_interpolate.py#L457

One  `np.interp`'s  feature to be careful about is its `left` and `right` arguments. Naively they are basically, `out[x > x[-1]] = right; out[x < x[0]] = left`, so need to take care of immutable arrays.

**Comment by lucascolley** [comment]
This sounds like a perfect candidate for a Python implementation in array-api-extra, then, for as long as https://scipy.github.io/devdocs/tutorial/interpolate/1D.html#piecewise-linear-interpolation continues to direct to NumPy.

**Comment by lucascolley** [comment]
next question: would `interp` be the best name @ev-br ? Or is there anything else that is more clear and doesn't clash with existing (semantically-different) functions?

**Comment by ev-br** [comment]
IMO, yes.

`xpx.pad` is a drop-in replacement for `np.pad`; `xpx.kron` is a drop-in replacement for `np.kron`, `xpx.interp` is a drop-in replacement for `np.interp`.

The ecosystem is already too crowded with variant functions for linear interpolation, which are almost but not completely unlike each other.

**Comment by ev-br** [comment]
The lack of `np.interp` showed up in https://github.com/scipy/scipy/pull/25246/

Quick check for availability in major array API librarues: 
- [CuPy has cp.interp](https://docs.cupy.dev/en/stable/reference/generated/cupy.interp.html), 
- [JAX has jax.numpy.interp](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.interp.html); 
- dask does not have it; workarounds exist, per  https://github.com/dask/dask/issues/6474 ;
- PyTorch does not have it, so will need to fall back to a "generic" implementation; 

For a "generic" version, can reuse either parts of `scipy.interpolate.interp1d` linked above, or the JAX implementation, https://github.com/jax-ml/jax/blob/main/jax/_src/numpy/lax_numpy.py#L2590-L2649

**Comment by prady0t** [comment]
Jax implementation would be a better fit for the extra module, as it follows the numpy API more closely. There are a few things we can discuss that could be (or not be) part of the standard.
- Jax accepts String values too for left and right to specify whether to extrapolate for points lying outside. SciPy uses `kind` (string) to specify the kind of interpolation. I feel we should just stick to the numpy behaviour and not take a string as an input at all; this would be restrictive, but much cleaner for a generic implementation.
- Assume that xp (here xp is the x-axis inputs) is sorted. SciPy gives an option to give unsorted input by setting `assume_sorted=False`, but Jax assumes sorted.

There are many more quirks in SciPy's implementation, which is definitely overkill. Following Jax's behaviour is much more helpful for the standard. The only place where Jax deviates (I feel) is the ability to extrapolate. We can discuees weather we want to keep it or not. I'll be happy to open a PR.

**Comment by ev-br** [comment]
<s>array-api-strict</s>`array-api-extra` should only implement what is available in numpy.interp. additional features of scipy.interpolate.interp1d should stay out.
